StyleCop.Analyzers/StyleCop.Analyzers/Helpers/DocumentBasedFixAllProvider.cs

```diff
@@ -42,21 +45,24 @@ internal abstract class DocumentBasedFixAllProvider : FixAllProvider
             }
         }
 
-        protected abstract Task<SyntaxNode> FixAllInDocumentAsync(FixAllContext fixAllContext, Document document);
+        protected abstract Task<SyntaxNode> FixAllInDocumentAsync(FixAllContext fixAllContext, Document document, ImmutableArray<Diagnostic> diagnostics);
 
         private async Task<Solution> GetProjectFixesAsync(FixAllContext fixAllContext, Project project)
         {
             Solution solution = project.Solution;
-            var oldDocuments = project.Documents.ToImmutableArray();
-            List<Task<SyntaxNode>> newDocuments = new List<Task<SyntaxNode>>(oldDocuments.Length);
-            foreach (var document in oldDocuments)
+            List<Document> oldDocuments = new List<Document>();
+            List<Task<SyntaxNode>> newDocumentRoots = new List<Task<SyntaxNode>>();
+            var diagnostics = await fixAllContext.GetProjectDiagnosticsAsync(project).ConfigureAwait(false);
+            var lookup = diagnostics.GroupBy(x => project.GetDocument(x.Location.SourceTree)).ToArray();
```

Comment:
Is `ToArray()` needed?  I don't know the internals of `project.GetDocument` but I would assume that avoiding an allocation here would be possible.